update twoliter and buildsys to put RPMs into per-package sub-directories

webern opened the pull request:
**Issue number:**

Related #185
Related #73
Builds on #198

**Description of changes:**

In order to construct kits containing the correct RPMs, we need to stop using a flat `build/rpms` structure. Instead we need to put each RPM into a sub-directory. Then we can use the package-name to find the right RPMs to build repos with the correct RPMs on-the-fly when building packages, kits and variants.

**Testing done:**

- [x] Build a Bottlerocket Variant
- [x] unit tests are passing
- [x] Built a Bottlerocket variant from develop (flat RPM structure) then tried to do an incremental build without the `twoliter: error if old rpm structure is discovered` commit. Buildsys failed as expected with `No matching package to install: 'bottlerocket-glibc-devel'`
- [x] Tried again with `twoliter: error if old rpm structure is discovered` , got the helpful message: `PLEASE RUN: cargo make clean`
- [x] Tried again after a `cargo make clean` and the build succeeded

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

Commit:
twoliter: embed rpm2kit

Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
twoliter: update Dockerfile for per-package rpm directories

Update the buildsys Dockerfile so that it works with RPM packages being
outputted to sub-directories, i.e. build/package-name/

Co-authored-by: Matt Briggs <brigmatt@amazon.com>
Co-authored-by: Ben Cressey <bcressey@amazon.com>
Signed-off-by: Ben Cressey <bcressey@amazon.com>

Commit:
buildsys: use per-package output directory

Place output RPMs into a subdirectory using the package-name.

webern committed:
buildsys: move package name override to manifest info

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/43ed803ea9183599a333687c97dd9fd3e5b5026e..3ded669d9a51d82e304a4555c75c71a2c6c1a5bf

Clippy Fix (boring)

bcressey commented on tools/buildsys/src/builder.rs:
```diff
@@ -247,11 +247,13 @@ impl DockerBuild {
-            args.cargo_package_name
-        };
+        let package = manifest.info().package_name();
+        let per_package_dir = format!(
+            "{}/{}",
+            args.packages_dir.display(),
+            manifest.info().manifest_name()
```
Are we able to use `package` here?
```suggestion
        let per_package_dir = format!(
            "{}/{}",
            args.packages_dir.display(),
            package,
```

bcressey commented on twoliter/embedded/Dockerfile.dockerignore:
```diff
@@ -3,9 +3,10 @@
 /build/*
 !/build/rpms/
 /build/rpms/*
-!/build/rpms/*.rpm
-/build/rpms/*-debuginfo-*.rpm
-/build/rpms/*-debugsource-*.rpm
+!/build/rpms/*/*.rpm
+/build/rpms/*/*-debuginfo-*.rpm
+/build/rpms/*/*-debugsource-*.rpm
+!/build/kits/
```
Maybe keep the earlier exclusions for now, with a comment that we can deprecate them eventually. Otherwise the build context will get very large for anyone that has existing files in `build/rpms` until they clean their checkout.
```suggestion
!/build/rpms/*.rpm
/build/rpms/*-debuginfo-*.rpm
/build/rpms/*-debugsource-*.rpm
!/build/rpms/*/*.rpm
/build/rpms/*/*-debuginfo-*.rpm
/build/rpms/*/*-debugsource-*.rpm
!/build/kits/
```

webern replied:
I've committed this suggestion, but don't we actually want to ignore `/build/rpms/*.rpm` to solve that problem?

bcressey commented on twoliter/embedded/rpm2kit:
```diff
@@ -0,0 +1,38 @@
+   esac
+done
+
+# HACK: because we're not passing KIT from buildsys
+export KIT="${VARIANT}"
```
```suggestion
```

bcressey commented on twoliter/embedded/rpm2kit:
```diff
@@ -0,0 +1,38 @@
+# HACK: because we're not passing KIT from buildsys
+export KIT="${VARIANT}"
+
+# FIXME: should this be a versioned directory?
+# - needs to change whenever the kit changes?
+# - needs to be predictable for publishing to work?
+# - needs to avoid ambiguity if we have other local builds of the same kit?
```
If we're emitting the right `cargo:rerun-if-changed` and `cargo:rerun-if-env-changed` directives in `buildsys` then we can drop this. Just need to make sure https://github.com/bottlerocket-os/twoliter/pull/198#discussion_r1594576970 is addressed.

webern replied:
I think the answer is as simple as `rm -rf "${KIT_DIR}"` since, if this script is running, then we know we are re-creating the kit. That's what I've pushed.

bcressey reviewed (approved):
Looks good, just some minor notes.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/6eaabe9a12d433f94b32747ba50ab05d2ee4c271..2108263fb3c1a34e932451132cc45e808fe0bb46

Rebase over #214

webern committed:
twoliter: error if old rpm structure is discovered

Add a check to Makefile.toml to see if the user has rpms in the
build/rpms directory. If so, this is likely a problem because the new
version of Buildsys will expect per-package subdirectories, and Cargo
will not know it needs to rebuild.

Raise an error that informs the user they need to run cargo make clean.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/3183c93da746776a9605b2008f2f036b1c22ba6c..b9d7639d19fe35c0ba44dd98f13782a1b30f8356

Add a check to Makefile.toml to tell the user when they need to run `cargo make clean`.

webern commented:
@bcressey this is ready for another review after adding the flat RPMs dir check.

webern commented:
> https://github.com/bottlerocket-os/twoliter/compare/b9d7639d19fe35c0ba44dd98f13782a1b30f8356..737ca5d2cca82b22a56e4647c037b72fbc2c8131

Rebase to fix Makefile.toml conflict

sam-berning commented on twoliter/embedded/rpm2kit:
```diff
@@ -0,0 +1,32 @@
+done
+
+createrepo_c "${KIT_DIR}"
+dnf --disablerepo '*' --repofrompath "kit,file:///${KIT_DIR}" repoquery --all
```
just for my understanding, what does this do? is it just disabling the kit dnf repo by default so that it has to be explicitly enabled in the variant build?
